### PR TITLE
Add support for flashing APM32 via DFU

### PR DESCRIPTION
**Pull request by awkannan**
## Description

Add support for flashing APM32 via DFU.
These are pin compatible with STM32.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

### CODE REVIEW

File: osx/qmk_toolbox/Flashing.m [hunk 1]
```diff
@@ -162,6 +164,14 @@ - (void)flashSTM32DFUWithFile:(NSString *)file {
     }
 }
 
+- (void)flashAPM32DFUWithFile:(NSString *)file {
+    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
+        [self runProcess:@"dfu-util" withArgs:@[@"-a", @"0", @"-d", @"314b:0106", @"-s", @"0x8000000:leave", @"-D", file]];
```
Comment (fauxpark) [inline osx/qmk_toolbox/Flashing.m:169]:
```suggestion
        [self runProcess:@"dfu-util" withArgs:@[@"-a", @"0", @"-d", @"314B:0106", @"-s", @"0x8000000:leave", @"-D", file]];
```

File: windows/QMK Toolbox/Flashing.cs [hunk 2]
```diff
@@ -28,6 +28,7 @@ public enum Chipset
         UsbTiny,
         BootloadHid,
         AtmelSamBa,
+        Apm32,
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:31]:
```suggestion
        Apm32Dfu,

```

File: windows/QMK Toolbox/Flashing.cs [hunk 3]
```diff
@@ -143,6 +144,8 @@ public void Flash(string mcu, string file)
                 FlashHalfkay(mcu, file);
             if (Usb.CanFlash(Chipset.Stm32))
                 FlashStm32(mcu, file);
+            if (Usb.CanFlash(Chipset.Apm32))
+                FlashApm32(mcu, file);
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:148]:
```suggestion
            if (Usb.CanFlash(Chipset.Apm32Dfu))
                FlashApm32Dfu(mcu, file);

```

File: windows/QMK Toolbox/Flashing.cs [hunk 4]
```diff
@@ -230,6 +233,15 @@ private void FlashStm32(string mcu, string file)
             }
         }
 
+        private void FlashApm32(string mcu, string file)
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:236]:
```suggestion
        private void FlashApm32Dfu(string mcu, string file)

```

File: windows/QMK Toolbox/Flashing.cs [hunk 5]
```diff
@@ -230,6 +233,15 @@ private void FlashStm32(string mcu, string file)
             }
         }
 
+        private void FlashApm32(string mcu, string file)
+        {
+            if (Path.GetExtension(file)?.ToLower() == ".bin") {
+                RunProcess("dfu-util.exe", $"-a 0 -d 314b:0106 -s 0x08000000:leave -D \"{file}\"");
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:239]:
```suggestion
                RunProcess("dfu-util.exe", $"-a 0 -d 314B:0106 -s 0x08000000:leave -D \"{file}\"");

```

File: windows/QMK Toolbox/USB.cs [hunk 6]
```diff
@@ -129,6 +129,11 @@ public bool DetectBootloader(ManagementBaseObject instance, bool connected = tru
                 deviceName = "STM32 DFU";
                 _devicesAvailable[(int)Chipset.Stm32] += connected ? 1 : -1;
             }
+            else if (MatchVidPid(deviceId, 0x314b, 0x0106)) // APM32 DFU
```
Comment (fauxpark) [inline windows/QMK Toolbox/USB.cs:132]:
```suggestion
            else if (MatchVidPid(deviceId, 0x314B, 0x0106)) // APM32 DFU
```

File: windows/QMK Toolbox/USB.cs [hunk 7]
```diff
@@ -129,6 +129,11 @@ public bool DetectBootloader(ManagementBaseObject instance, bool connected = tru
...
+            {
+                deviceName = "APM32 DFU";
+                _devicesAvailable[(int)Chipset.Apm32] += connected ? 1 : -1;
```
Comment (fauxpark) [inline windows/QMK Toolbox/USB.cs:135]:
```suggestion
                _devicesAvailable[(int)Chipset.Apm32Dfu] += connected ? 1 : -1;
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by fauxpark** [comment]
So I just discovered something odd... the dfu-util command for the STM32 bootloader specifies the wrong VID (`0482`) in the macOS code, and yet it is able to flash my Proton-C just fine. So it seems like the `-d` flag is being ignored. Maybe we don't need a separate flasher for APM32? Or is dfu-util simply reading the suffix instead?

If so... I have a plan to add `BOOTLOADER = stm32-dfu` into QMK, which currently would only set the dfu-util and dfu-suffix arguments, but this coupled with an `apm32-dfu` option would probably allow us to remove the `-d` flag altogether and just have the STM32 flasher handle both in a much more generic fashion.

**Comment by awkannan** [comment]
@fauxpark I've only tested on Windows, but the -d does apply.

WIthout the -d, dfu-util will flash any dfu device that's connected, and skips the suffix check.

I can't speak to why the mac os x one is working. Are you sure that there is a suffix on the bin file?

**Comment by fauxpark** [comment]
Just from a quick look at the code, it doesn't seem like a lack of `-d` skips the suffix check. In fact, it appears dfu-util specifically inspects the suffix to determine which device to flash:

```c
		dfu_load_file(&file, MAYBE_SUFFIX, MAYBE_PREFIX);
		/* If the user didn't specify product and/or vendor IDs to match,
		 * use any IDs from the file suffix for device matching */
		if (match_vendor < 0 && file.idVendor != 0xffff) {
			match_vendor = file.idVendor;
			printf("Match vendor ID from file: %04x\n", match_vendor);
		}
		if (match_product < 0 && file.idProduct != 0xffff) {
			match_product = file.idProduct;
			printf("Match product ID from file: %04x\n", match_product);
		}

```
https://sourceforge.net/p/dfu-util/dfu-util/ci/master/tree/src/main.c#l368

If there is no suffix (recognised here as `idVendor`/`idProduct` set to `0xFFFF`), then `-d` would be needed (and dfu-util spits out a warning afterwards that it is missing).

Thus if we are able to add the correct suffix during compilation, through `BOOTLOADER`, there should be no need for the duplicate flashing method, at the very least. The detection stuff could all stay, so that the Toolbox correctly reports APM32 rather than STM32, but they would both call the same method.

**Comment by awkannan** [comment]
I think the main issue here, though, is each bin file can only have a single suffix. In this instance, the same firmware can work with both STM32 and APM32.

Testing locally, with a bin file with no suffix, and a dfu-util command with no `-d` specified, it'll flash both STM32 and APM32 just fine, thought it may throw a warning.

But if someone has both an STM32 and APM32 device plugged in and in DFU mode, it'll just pick one at random iirc.
Which is why I think no suffix + -d in toolbox is probably the best way.

Then again, this is the same problem if someone has 2 STM32 devices plugged in in DFU mode... I can see it both ways.

**Comment by fauxpark** [comment]
I see this primarily as a problem with how the Toolbox is currently designed - if you have multiple bootloader devices (of any kind) connected, it will attempt to flash all of them. Eventually there will be some refactoring and UI updates to allow users to select which device to flash. But for the time being, I guess it is kind of an unspoken rule that you should only flash one board at a time.

IMO the correct suffix should be set at compile time, so that dfu-util does not print the warning message - this might confuse users, thinking that something went wrong. I've seen this happen before with other things, not just the Toolbox but also QMK compilation itself.

**Comment by awkannan** [comment]
> IMO the correct suffix should be set at compile time, so that dfu-util does not print the warning message - this might confuse users, thinking that something went wrong. I've seen this happen before with other things, not just the Toolbox but also QMK compilation itself.

The issue that for a given firmware, they can be multiple correct suffixes. STM32F072CBT6 firmware works on APM32F072CBT6, and vice versa but they have different suffixes.

We can implement logic in QMK Toolbox to strip/re-add the suffix if it's `0483:DF11` and we are flashing APM32. This was the first approach I suggested but @drashna suggested just clearing out the suffix.

**Comment by fauxpark** [comment]
Personally I would prefer to stay ahead of the warning message, because at some point dfu-util is going to update, and this will not work anymore. If that means splitting boards up into STM32 and APM32 "revisions", so be it. (you would only need to set the `DFU_ARGS`/`DFU_SUFFIX_ARGS`, or potentially `BOOTLOADER`, though)

See also the mdloader drama, where Drop substituted their CTRL/ALTs' MCUs for the same one but with more flash, resulting in nothing being detected by mdloader because the chip ID was different.

**Comment by sigprof** [comment]
Apparently there is an option to append the DFU suffix with wildcard values (0xffff) for `idVendor` and `idProduct`, so that dfu-util won't complain about the missing suffix, but also won't perform the extra VID:PID check.

**Comment by awkannan** [comment]
Given that 0xFFFF 0xFFFF works as a "wildcard" suffix, I think this is still the way to go.

File: windows/QMK Toolbox/Flashing.cs [hunk 8]
```diff
@@ -230,6 +235,17 @@ private void FlashStm32(string mcu, string file)
             }
         }
 
+        private void FlashApm32Dfu(string mcu, string file)
+
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:239]:
```suggestion

```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Could you also add APM32 to the supported bootloaders list (ie `ARM DFU (STM32, APM32, Kiibohd)`) in readme.md, AppDelegate.m, main.m, MainWindow.cs and Program.cs?

Also, what's the "Bus reported device description" for this bootloader in Device Manager? I'd like to know so I can list it in the Zadig docs:
image

**Comment by awkannan** [comment]
@fauxpark I'll rebase and make those suggested changes tonight. Thanks!

**Review by fauxpark — APPROVED** [review]

**Comment by tzarc** [comment]
@fauxpark do we need to add the VID/PID pair to `drivers.txt`?

**Comment by fauxpark** [comment]
Oops, yeah.

**Comment by awkannan** [comment]
Added to drivers.txt

**Review by tzarc — APPROVED** [review]

**Pull request merged**